ocaml/sdk-gen/go/gen_go_helper.ml

```diff
@@ -45,35 +45,57 @@ let generate_file rendered output_file =
     (fun () -> output_string out_chan rendered)
     ~finally:(fun () -> close_out out_chan)
 
+module EnumSet = Set.Make (struct
+  type t = string * (string * string) list
+
+  let compare (x0, _y0) (x1, _y1) = String.compare x0 x1
+end)
+
 module Json = struct
-  let rec string_of_ty ty =
+  let rec string_of_ty_with_enums ty =
     match ty with
     | SecretString | String ->
-        "string"
+        ("string", [])
     | Int ->
-        "int"
+        ("int", [])
     | Float ->
-        "float64"
+        ("float64", [])
     | Bool ->
-        "bool"
+        ("bool", [])
     | DateTime ->
-        "time.Time"
-    | Enum (name, _kv) ->
-        snake_to_camel name
+        ("time.Time", [])
+    | Enum (name, kv) ->
+        let name = snake_to_camel name in
+        (name, [(name, kv)])
     | Set ty ->
-        "[]" ^ string_of_ty ty
+        let s, e = string_of_ty_with_enums ty in
+        ("[]" ^ s, e)
     | Map (ty1, ty2) ->
-        let s1 = string_of_ty ty1 in
-        let s2 = string_of_ty ty2 in
-        "map[" ^ s1 ^ "]" ^ s2
+        let s1, e1 = string_of_ty_with_enums ty1 in
+        let s2, e2 = string_of_ty_with_enums ty2 in
+        let ty = "map[" ^ s1 ^ "]" ^ s2 in
+        (ty, e1 @ e2)
     | Ref r ->
-        snake_to_camel r ^ "Ref"
+        (snake_to_camel r ^ "Ref", [])
     | Record r ->
-        snake_to_camel r ^ "Record"
+        (snake_to_camel r ^ "Record", [])
     | Option ty ->
-        string_of_ty ty
+        string_of_ty_with_enums ty
```

Comment:
The list here should be replaced with empty maps, instead of concatenating the list, the unions of the maps have to be calculated:

```suggestion
  let rec string_of_ty_with_enums ty =
    match ty with
    | SecretString | String ->
        ("string", StringMap.empty)
    | Int ->
        ("int", StringMap.empty)
    | Float ->
        ("float64", StringMap.empty)
    | Bool ->
        ("bool", StringMap.empty)
    | DateTime ->
        ("time.Time", StringMap.empty)
    | Enum (name, kv) ->
        let name = snake_to_camel name in
        (name, StringMap.singleton name kv)
    | Set ty ->
        let s, e = string_of_ty_with_enums ty in
        ("[]" ^ s, e)
    | Map (ty1, ty2) ->
        let s1, e1 = string_of_ty_with_enums ty1 in
        let s2, e2 = string_of_ty_with_enums ty2 in
        let ty = "map[" ^ s1 ^ "]" ^ s2 in
        (ty, StringMap.union choose_enum e1 e2)
    | Ref r ->
        (snake_to_camel r ^ "Ref", StringMap.empty)
    | Record r ->
        (snake_to_camel r ^ "Record", StringMap.empty)
    | Option ty ->
        string_of_ty_with_enums ty
```

In the preceding comment I added the choose_enum. Since the datamodel is consistent I don't think it matters which one of the two values is picked, I picked the left one.

Other functions need to be transformed in similar ways